perceiver/data/audio/symbolic.py

```diff
@@ -0,0 +1,216 @@
+import functools
+import multiprocessing
+import os
+import random
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import numpy as np
+import pytorch_lightning as pl
+import torch
+import torch.nn.functional as F
+from torch.utils.data import DataLoader
+
+from perceiver.data.audio.midi_processor import encode_midi_files
+
+
+class SymbolicAudioDataModule(pl.LightningDataModule):
+    _EXAMPLE_SEPARATOR_INPUT_ID = -1
+    _PAD_INPUT_ID = 388
+    _VOCAB_SIZE = 389
+
+    def __init__(
+        self,
+        dataset_dir: str,
+        max_seq_len: int,
+        min_seq_len: Optional[int] = None,
+        padding_side: Optional[str] = "left",
+        batch_size: int = 16,
+        num_workers: int = 1,
+        preproc_workers: int = multiprocessing.cpu_count(),
+        pin_memory: bool = True,
+    ):
+        super().__init__()
+        if min_seq_len is not None and not (0 < min_seq_len < max_seq_len):
+            raise ValueError(
+                "Invalid data configuration supplied. "
+                "Parameter 'min_seq_len' must adhere to 0 < min_seq_len < max_seq_len."
+            )
+
+        self.save_hyperparameters()
+
+        self._collator = SymbolicAudioCollator(
+            max_seq_len=self.hparams.max_seq_len + 1,
+            pad_token=self._PAD_INPUT_ID,
+            padding_side=self.hparams.padding_side,
+        )
+        self._ds_train = None
+        self._ds_valid = None
+
+    @property
+    def vocab_size(self):
+        return self._VOCAB_SIZE
+
+    @property
+    def max_seq_len(self):
+        return self.hparams.max_seq_len
+
+    @property
+    @functools.lru_cache(maxsize=1)
+    def preproc_dir(self) -> Path:
+        return Path(self.hparams.dataset_dir) / "preproc" / self.preproc_dir_hash()
+
+    def preproc_dir_hash(self) -> str:
```

Comment:
I think we can remove this and the implementations in subclasses. It is safe to assume that different `dataset_uri`s will always host the same dataset file.